lib/functional/functional_gen.nit

```diff
@@ -0,0 +1,166 @@
+module functional_gen
+
+# Writer who adds new line at the end of each `writeln`.
+# Supports tabs indentation
+abstract class WriterLn
+        super Writer
+
+        # Number of tabulation to write before
+        var tabs = 0
+
+        # Adds `\n` at the end
+        fun writeln(s: Text)
+        do
+                for t in [0..tabs[ do write("\t")
+                write(s + "\n")
+        end
+end
+
+# A File writer who can add a new line at the end of each `writeln`
+class FileWriterLn
+        super FileWriter
+        super WriterLn
+end
+
+# Generates of an array of formal type as strings.
+# The size of the array equals the arity of the class.
+fun gen_generics(nargs: Int): Array[String]
+do
+        var args = new Array[String]
+        for i in [0..nargs[ do
+                args.push("A" + i.to_s)
+        end
+        return args
+end
+
+
+# Base class to writes Nit's functional classes as text.
+class FunTypeWriter
+
+        # Name of the class
+        var name: String
+
+        # Super classes as text
+        var supers: SequenceRead[String]
+
+        # The arity of the class
+        var nb_formal_types: Int
+
+        # Adds another formal type to express a return type
+        var with_return = false
+
+        # Annotation to add after `call`
+        var annotation = "is abstract"
+
+        # Writes a class from provided attributes
+        fun write(writer: WriterLn)
+        do
+                var i = nb_formal_types
+                var generics = gen_generics(i)
+                if with_return then
+                        generics.push("RESULT")
+                end
+                writer.write("interface {name}{i}")
+                writer.writeln("[" + generics.join(",") + "]")
+                writer.tabs += 1
+                for zuper in supers do
+                        writer.writeln("super {zuper}")
+                end
+                writer.tabs -= 1
+                var output = ""
+                if with_return then
+                        output = generics.pop
+                end
+                var params = new Array[String]
+                for g in generics do
+                        params.push(g.to_lower + ": " + g)
+                end
+
+                var signature = params.join(",")
+                if i == 0 then
+                        writer.write("\tfun call")
+                else
+                        writer.write("\tfun call(" + signature + ")")
+                end
+                if with_return then
+                        writer.write(": {output}")
+                end
+                writer.writeln(" {annotation}")
+                writer.writeln("end")
+        end
+
+end
+
+# Writes all functional types
+fun generate_functypes(n: Int, writer: WriterLn)
+do
+        writer.writeln("""
+# This file is part of NIT ( http://www.nitlanguage.org ).
+#
+# Copyright 2019 Louis-Vincent Boudreault <lv.boudreault95@gmail.com>
```

Comment:
Not sure this is a good idea to generate that copyright. What if **I** use the generator or I do it in **2121**.

As you can see I hate these comments :D